### PR TITLE
Replace NULL with C++11 nullptr

### LINKED ISSUE #1768: Replace NULL by nullptr?
Idea from the C++ Hackathon 

So far we have 639 occurences of `NULL` in our most recent dev branch. 
Should we replace many/all of them by C++11 `nullptr`? 
[url](http://en.cppreference.com/w/cpp/language/nullptr)

### PR DESCRIPTION
**Pull request by ax3l**
Close  #1768 via:

```bash
grep -R "NULL" src/ | awk -F: '{print $1}' | sort -u | \
  xargs -n1 -P1 -I{} sed -i '/_NULL/b; s/NULL/nullptr/g' {}

GIT_AUTHOR_NAME="Tools" GIT_AUTHOR_EMAIL="picongpu@hzdr.de" git commit -a -S
```

### CODE REVIEW

File: src/picongpu/include/plugins/adios/ADIOSWriter.def [hunk 1]
```diff
@@ -74,7 +74,7 @@ namespace po = boost::program_options;
     }                                                                         \
 }
 
-#define ADIOS_CMD_EXPECT_NONNULL(_cmd)                                        \
+#define ADIOS_CMD_EXPECT_NONnullptr(_cmd)                                        \
```
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/adios/ADIOSWriter.def]: please revert this or change it to `ADIOS_CMD_EXPECT_NON_NULLPTR`

File: src/picongpu/include/plugins/adios/NDScalars.hpp [hunk 2]
```diff
@@ -118,13 +118,13 @@ struct ReadNDScalars
...
                 const std::string& name, T_Scalar* value,
-                const std::string& attrName = "", T_Attribute* attribute = NULL)
+                const std::string& attrName = "", T_Attribute* attribute = nullptr)
     {
         log<picLog::INPUT_OUTPUT> ("ADIOS: read %1%D scalars: %2%") % simDim % name;
         std::string datasetName = params.adiosBasePath + name;
 
         ADIOS_VARINFO* varInfo;
-        ADIOS_CMD_EXPECT_NONNULL( varInfo = adios_inq_var(params.fp, datasetName.c_str()) );
+        ADIOS_CMD_EXPECT_NONnullptr( varInfo = adios_inq_var(params.fp, datasetName.c_str()) );
```
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: wrong substitution

File: src/picongpu/include/plugins/adios/ADIOSWriter.def (hunk 1 above)
Comment (ax3l) [inline src/picongpu/include/plugins/adios/ADIOSWriter.def]: huh, well spotted. dang, I tried to look careful!

File: src/picongpu/include/plugins/adios/NDScalars.hpp (hunk 2 above)
Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: ty!

**Pull request merged**